src/Language/ASKEE/ExprTransform.hs

```diff
@@ -102,6 +102,63 @@ inlineLets' lets m = runIdentity (transformModelExprs go m)
             Nothing -> pure e
         _ -> pure e
 
+-- | Rename variables in an expression via the provided function
+renameExprVarsWith :: (Text -> Text) -> Expr.Expr -> Expr.Expr
+renameExprVarsWith r e = runIdentity $ transformExpr go e
+  where
+    go ex =
+      case ex of
+        Expr.Var v -> pure $ Expr.Var (r v)
+        _          -> pure ex
+
+-- | Rename variables in an event via the provided function
+--
+-- Note: this will also rename variables that appear on event effect LHSs
+renameEventVarsWith :: (Text -> Text) -> Syntax.Event -> Syntax.Event
+renameEventVarsWith r = runIdentity . modifyEventVars (pure . r)
+  where
+    modifyEventVars varT evt =
+      do  when'   <- expr `traverse` Syntax.eventWhen evt
+          rate'   <- expr (Syntax.eventRate evt)
+          effect' <- transformStmt `traverse` Syntax.eventEffect evt
+          name' <- varT (Syntax.eventName evt)
+          pure $ evt  { Syntax.eventWhen = when'
+                      , Syntax.eventRate = rate'
+                      , Syntax.eventEffect = effect'
+                      , Syntax.eventName = name'
+                      }
+      where
+        exprT e =
+          case e of
+            Expr.Var v -> Expr.Var <$> varT v
+            _     -> pure e
+
+        transformStmt (n, v) = 
+          do  n' <- varT n
+              v' <- expr v
+              pure (n', v')
+
+        expr = transformExpr exprT
+
+
+renameModelVarsWith :: (Text -> Text) -> Syntax.Model -> Syntax.Model
+renameModelVarsWith r = runIdentity . modifyModelVars
+  where
+    modifyModelVars mdl =
+      do  let decls' = map transformDecl (Syntax.modelDecls mdl)
+          let events' = map (renameEventVarsWith r) (Syntax.modelEvents mdl)
+          pure mdl 
```

Comment:
Is there any point in using `pure` here? As far as I can tell, `modifyModelVars` isn't doing anything effectful, and thus shouldn't require `Applicative`.

Similar questions apply to `renameExprVarsWith` and `renameEventVarsWith`.

Reply:
In that function you're right, there's no need to go monadic/applicative. The other functions you name rely on `transformExpr :: Monad m => (Expr -> m Expr) -> Expr -> m Expr`, and so need to encapsulate that with an `Identity` computation.